perfkitbenchmarker/virtual_machine.py

```diff
@@ -34,6 +34,38 @@
 FLAGS = flags.FLAGS
 DEFAULT_USERNAME = 'perfkit'
 
+_VM_SPEC_REGISTRY = {}
+_VM_REGISTRY = {}
+
+
+def GetVmSpecClass(cloud):
+  """Returns the VmSpec class corresponding to 'cloud'."""
+  return _VM_SPEC_REGISTRY.get(cloud, BaseVmSpec)
+
+
+def GetVmClass(cloud, os_type):
+  """Returns the VM class corresponding to 'cloud' and 'os_type'."""
+  return _VM_REGISTRY.get((cloud, os_type))
+
+
+class AutoRegisterVmSpecMeta(type):
+  """Metaclass which allows VmSpecs to automatically be registered."""
+
+  def __init__(cls, name, bases, dct):
+    if cls.CLOUD:
+      _VM_SPEC_REGISTRY[cls.CLOUD] = cls
+    super(AutoRegisterVmSpecMeta, cls).__init__(name, bases, dct)
+
+
+class AutoRegisterVmMeta(abc.ABCMeta):
+  """Metaclass which allows VMs to automatically be registered."""
+
+  def __init__(cls, name, bases, dct):
+    if (hasattr(cls, 'CLOUD') and cls.CLOUD and
```

Comment:
Same comment as for disk specs - should it be an error *not* to have CLOUD and OS_TYPE attributes?